New pgo format for crossgen2

davidwrighton opened the pull request:
Define new experimental pgo file format, and adjust crossgen2 to be able to consume it
- This format is designed to represent data from an entire scenario run, not just data from one module
- And the format is version resilient such that data shouldn't grow stale if tokens change
- With this change, an arbitrary number of pgo scenarios may be fed into the compiler for future processing.
- The new format is represented as a compressed il dll and leverages the format to provide version resiliency.

@dotnet/crossgen-contrib 

In addition this change contains fixes found while attempting to use our existing profile guided functionality.
- IBC processing could fail to find types/methods while parsing data (and as IBC data is fundamentally unreliable, this was common)
- System.__Canon was loadable as both an `EcmaType` as well as a `CanonType`
- The intrinsic for replacing Activator.CreateInstance with Activator.CreateDefaultInstance could produce a debug assert when used with a structure instantiated over a reference type

davidwrighton committed:
Adjust compilation model to support trace based profile guided optimization data in a new format
- Add parser for new format
- Adjust program startup and options parsing to make it simpler for the profiler guided logic to be aware of version bubbles
- Use CompilationModuleGroup and version bubble in parsing and handling of profile data

davidwrighton committed:
Handle malformed and incorrect IBC data

davidwrighton committed:
Fixup bad merge

davidwrighton committed:
Fixes to make standalone build of csc compile with a profile

jkotas commented on src/coreclr/src/tools/Common/TypeSystem/Ecma/EcmaModule.cs:
```diff
@@ -90,7 +90,15 @@ protected override IEntityHandleObject CreateValueFromKey(EntityHandle handle)
-                        item = new EcmaType(_module, (TypeDefinitionHandle)handle);
+                        var ecmaType = new EcmaType(_module, (TypeDefinitionHandle)handle);
+                        item = ecmaType;
+                        if (_module == _module.Context.CanonTypesModule)
```
`_module.Context` is virtual property. It would be nice to fetch it just once into a local.

davidwrighton commented on src/coreclr/src/tools/Common/TypeSystem/Ecma/EcmaModule.cs:
```diff
@@ -90,7 +90,15 @@ protected override IEntityHandleObject CreateValueFromKey(EntityHandle handle)
+                        item = ecmaType;
+                        if (_module == _module.Context.CanonTypesModule)
+                        {
+                            if (_module.Context.SupportsCanon && _module.Context.CanonType != null && (ecmaType.Name == _module.Context.CanonType.Name) && (ecmaType.Namespace == _module.Context.CanonType.Namespace))
```
I knew there was a reason we hadn't done this before. I'm going to try the Canon lives in a magical assembly approach, and that should clean this up. I wasn't particularly pleased by this part of my change, and how I kept finding new ways that it broke stuff.

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/IBC/MIbcProfileParser.cs:
```diff
@@ -0,0 +1,182 @@
+                }
+            }
+
+            EcmaModule mibcModule = tsc.GetMetadataOnlyModuleFromMemory(filename, peData);
```
Good point, it hadn't occurred to me that was a viable option, but it likely is, and that keeps this out of the binding discussion entirely. (It already doesn't participate in binding, but it fills up some data structures for no good reason.)

davidwrighton committed:
Remove unused new command line parameter "non-local-generics"
Revert changes around System.__Canon handling when loading from System.Private.CoreLib

davidwrighton committed:
Handle System.Private.Canon instances without polluting the general purpose type system

davidwrighton committed:
Revert changes to CompilerTypeSystemContext

MichalStrehovsky commented on src/coreclr/src/tools/Common/TypeSystem/Common/TypeSystemContext.cs:
```diff
@@ -66,7 +66,7 @@ public virtual ModuleDesc ResolveAssembly(AssemblyName name, bool throwIfNotFoun
         }
 
-        internal virtual ModuleDesc ResolveModule(IAssemblyDesc referencingModule, string fileName, bool throwIfNotFound = true)
+        ModuleDesc IModuleResolver.ResolveModule(IAssemblyDesc referencingModule, string fileName, bool throwIfNotFound)
```
ILVerify supports multi-module assemblies and overrides this. I think this will be a build break there. Maybe leaving it as non-explicit implementation is enough?

davidwrighton replied:
Fixed by putting the internal virtual method back and calling that from IModuleResolver.ResolveModule.

davidwrighton committed:
Prevent build break in ILVerify
